File: agent/local_customers.py
"""
Base de datos de clientes local por perfil.
"""
import json
from pathlib import Path
from agent.profiles import get_profile_data_dir


def _customers_file(profile_id: str) -> Path:
    return get_profile_data_dir(profile_id) / "customers.json"


def _load(profile_id: str) -> list:
    f = _customers_file(profile_id)
    return json.loads(f.read_text(encoding="utf-8")) if f.exists() else []


def _save(profile_id: str, customers: list):
    _customers_file(profile_id).write_text(
        json.dumps(customers, ensure_ascii=False, indent=2), encoding="utf-8"
    )


def list_customers(profile_id: str) -> list[dict]:
    return sorted(_load(profile_id), key=lambda x: x.get("name", "").lower())


def get_customer(profile_id: str, customer_id: str) -> dict | None:
    return next((c for c in _load(profile_id) if c.get("id") == customer_id), None)


def search_customers(profile_id: str, query: str) -> list[dict]:
    q = query.lower()
    return [
        c for c in _load(profile_id)
        if q in c.get("name", "").lower()
        or q in c.get("phone", "").lower()
        or q in c.get("email", "").lower()
    ]


def upsert_customer(profile_id: str, customer: dict) -> dict:
    customers = _load(profile_id)
    if not customer.get("id"):
        import uuid
        customer["id"] = str(uuid.uuid4())[:8]
    customers = [c for c in customers if c.get("id") != customer["id"]]
    customers.append(customer)
    _save(profile_id, customers)
    return customer


def delete_customer(profile_id: str, customer_id: str):
    customers = [c for c in _load(profile_id) if c.get("id") != customer_id]
    _save(profile_id, customers)


def import_from_sales(profile_id: str) -> int:
    """Importa clientes únicos desde el historial de ventas. Retorna cantidad agregada."""
    from agent.local_sales import list_invoices
    import uuid

    existing = _load(profile_id)
    existing_keys = {
        (c.get("name", "").strip().lower(), c.get("phone", "").strip())
        for c in existing
    }
    added = 0
    for inv in list_invoices(profile_id, limit=9999):
        name  = (inv.get("customer") or "").strip()
        phone = (inv.get("customer_phone") or "").strip()
        if not name:
            continue
        key = (name.lower(), phone)
        if key in existing_keys:
            continue
        existing_keys.add(key)
        existing.append({
            "id":      str(uuid.uuid4())[:8],
            "name":    name,
            "phone":   phone,
            "address": (inv.get("customer_address") or "").strip(),
            "rfc":     (inv.get("customer_rfc") or "").strip(),
        })
        added += 1
    if added:
        _save(profile_id, existing)
    return added

File: desktop_java/src/main/java/com/mfagent/service/LocalCustomers.java
package com.mfagent.service;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.nio.file.*;
import java.util.*;

/**
 * CRUD de clientes local por perfil.
 * Equivalente a agent/local_customers.py
 */
public class LocalCustomers {

    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private final String profileId;

    public LocalCustomers(String profileId) {
        this.profileId = profileId;
    }

    private Path file() {
        return ProfilePaths.dataDir(profileId).resolve("customers.json");
    }

    private List<Map<String, Object>> load() {
        if (!Files.exists(file())) return new ArrayList<>();
        try {
            return GSON.fromJson(Files.readString(file()),
                    new TypeToken<List<Map<String, Object>>>(){}.getType());
        } catch (Exception e) { return new ArrayList<>(); }
    }

    private void save(List<Map<String, Object>> customers) {
        try {
            Files.createDirectories(file().getParent());
            Files.writeString(file(), GSON.toJson(customers));
        } catch (Exception e) { System.err.println("[Customers] " + e.getMessage()); }
    }

    public List<Map<String, Object>> list() {
        List<Map<String, Object>> all = load();
        all.sort(Comparator.comparing(c -> c.getOrDefault("name", "").toString().toLowerCase()));
        return all;
    }

    public List<Map<String, Object>> search(String query) {
        String q = query.toLowerCase();
        return load().stream().filter(c ->
                c.getOrDefault("name",    "").toString().toLowerCase().contains(q) ||
                c.getOrDefault("phone",   "").toString().toLowerCase().contains(q) ||
                c.getOrDefault("address", "").toString().toLowerCase().contains(q)
        ).toList();
    }

    public Map<String, Object> upsert(Map<String, Object> customer) {
        List<Map<String, Object>> all = load();
        if (customer.get("id") == null || customer.get("id").toString().isEmpty())
            customer.put("id", UUID.randomUUID().toString().substring(0, 8));
        all.removeIf(c -> c.get("id").equals(customer.get("id")));
        all.add(customer);
        save(all);
        return customer;
    }

    public void delete(String id) {
        List<Map<String, Object>> all = load();
        all.removeIf(c -> id.equals(c.get("id")));
        save(all);
    }

    /** Importa clientes únicos desde el historial de ventas. Retorna cantidad agregada. */
    public int importFromSales() {
        List<Map<String, Object>> existing = load();
        Set<String> keys = new HashSet<>();
        for (Map<String, Object> c : existing)
            keys.add(c.getOrDefault("name","").toString().trim().toLowerCase()
                    + "|" + c.getOrDefault("phone","").toString().trim());

        int added = 0;
        for (var inv : new LocalSales(profileId).list(9999)) {
            String name  = inv.getCustomer()      != null ? inv.getCustomer().trim()         : "";
            String phone = inv.getCustomerPhone() != null ? inv.getCustomerPhone().trim()    : "";
            if (name.isEmpty()) continue;
            String key = name.toLowerCase() + "|" + phone;
            if (keys.contains(key)) continue;
            keys.add(key);
            Map<String, Object> c = new LinkedHashMap<>();
            c.put("id",      UUID.randomUUID().toString().substring(0, 8));
            c.put("name",    name);
            c.put("phone",   phone);
            c.put("address", inv.getCustomerAddress() != null ? inv.getCustomerAddress().trim() : "");
            c.put("rfc",     inv.getCustomerRfc()     != null ? inv.getCustomerRfc().trim()     : "");
            existing.add(c);
            added++;
        }
        if (added > 0) save(existing);
        return added;
    }
}
